Resources/public/js/router.js

```diff
@@ -1,12 +1,400 @@
+                        delete unusedParams[token[3]];
+                    }
+
+                    return;
+                }
+
+                throw new Error('The token type "' + token[0] + '" is not supported.');
+            });
+
+            if (url === '') {
+                url = '/';
+            }
+
+            route.hosttokens.forEach(function (token) {
+                var value = void 0;
+
+                if ('text' === token[0]) {
+                    host = token[1] + host;
+
+                    return;
+                }
+
+                if ('variable' === token[0]) {
+                    if (token[3] in params) {
+                        value = params[token[3]];
+                        delete unusedParams[token[3]];
+                    } else if (route.defaults && token[3] in route.defaults) {
+                        value = route.defaults[token[3]];
+                    }
+
+                    host = token[1] + value + host;
+                }
+            });
+
+            url = this.context_.base_url + url;
+            if (route.requirements && "_scheme" in route.requirements && this.getScheme() != route.requirements["_scheme"]) {
+                url = route.requirements["_scheme"] + "://" + (host || this.getHost()) + url;
+            } else if (host && this.getHost() !== host) {
+                url = this.getScheme() + "://" + host + url;
+            } else if (absolute === true) {
+                url = this.getScheme() + "://" + this.getHost() + url;
+            }
+
+            if (Object.keys(unusedParams).length > 0) {
+                (function () {
+                    var prefix = void 0;
+                    var queryParams = [];
+                    var add = function add(key, value) {
+                        // if value is a function then call it and assign it's return value as value
+                        value = typeof value === 'function' ? value() : value;
+
+                        // change null to empty string
+                        value = value === null ? '' : value;
+
+                        queryParams.push(encodeURIComponent(key) + '=' + encodeURIComponent(value));
+                    };
+
+                    for (prefix in unusedParams) {
+                        _this2.buildQueryParams(prefix, unusedParams[prefix], add);
+                    }
+
+                    url = url + '?' + queryParams.join('&').replace(/%20/g, '+');
+                })();
+            }
+
+            return url;
+        }
+    }], [{
+        key: 'getInstance',
+        value: function getInstance() {
+            return instance || new Router();
+        }
+
+        /**
+         * Configures the current Router instance with the provided data.
+         * @param {Object} data
+         */
+
+    }, {
+        key: 'setData',
+        value: function setData(data) {
+            var router = Router.getInstance();
+
+            router.setBaseUrl( /** @type {string} */data['base_url']);
```

Comment:
all these ``@type`` are useless once you don't use the Google closure compiler (we were running it in strict mode)